src/services/standardGoals.ts

```diff
@@ -0,0 +1,243 @@
+  // and have been told that is all we should expect)
+  if (requiresPrompts && !rootCauses) {
+    throw new Error('Root causes are required for this goal');
+  }
+
+  return { standard, requiresPrompts };
+}
+
+// This function will handle
+// - creating a new standard goal
+// - based on the design, this will not unsuspend or "restart" a closed goal
+// - creating new objectives if appropriate
+// - This will also serve to restart a standard goal (that has been closed)
+export async function newStandardGoal(
+  grantId: number,
+  standardGoalId: number,
+  objectives?: Array<IObjective>,
+  // todo: if we ever add more prompt responses, we will need to make this next param generic
+  rootCauses?: Array<string>,
+) {
+  const { standard, requiresPrompts } = await getStardard(standardGoalId, grantId, rootCauses);
+
+  if (standard.goals.length > 0) {
+    throw new Error('Standard goal has already been utilized');
+  }
+
+  const newGoal = await Goal.create({
+    status: GOAL_STATUS.NOT_STARTED,
+    name: standard.templateName,
+    grantId,
+    goalTemplateId: standard.id,
+    createdVia: 'rtr',
+  });
+
+  // a new goal does not require objectives, but may include them
+  if (objectives && objectives.length) {
+    await Objective.bulkCreate(objectives.map((objective) => ({
+      ...objective,
+      createdVia: 'rtr',
+      status: GOAL_STATUS.NOT_STARTED,
+      goalId: newGoal.id,
+    })));
+  }
+
+  const { prompts } = standard;
+
+  if (requiresPrompts) {
+    await GoalFieldResponse.create({
+      goalId: newGoal.id,
+      // again, we only foresee one possible prompt right now
+      goalTemplateFieldPromptId: prompts[0].id,
+      response: rootCauses,
+    });
+  }
+
+  return goalForRtr(
+    grantId,
+    standardGoalId,
+  );
+}
+
+// This function will handle
+// - editing a standard goal
+//
+export async function updateExistingStandardGoal(
+  grantId: number,
+  standardGoalId: number,
+  objectives: Array<IObjective>, // expect objectiveTemplateId here
+  rootCauses?: Array<string>,
+) {
+  const { standard, requiresPrompts } = await getStardard(standardGoalId, grantId, rootCauses);
+
+  if (standard.goals.length !== 1) {
+    throw new Error('Grant has not utilized this standard goal, or has more than one active standard goal');
+  }
+
+  const [goal] = standard.goals;
+  const { prompts } = standard;
+
+  // a new goal does not require objectives, but may include them
+  if (objectives.length) {
+    await Promise.all(objectives.map(async (objective) => {
+      if (objective.objectiveTemplateId) {
+        const existingObjective = await Objective.findOne({
+          where: {
+            goalId: goal.id,
+            objectiveTemplateId: objective.objectiveTemplateId,
+          },
+        });
+
+        if (existingObjective) {
+          return existingObjective.update({
+            title: objective.title,
+          });
+        }
+      }
+      return Objective.create({
```

Comment:
Should this create the objective template also? I ask because I dont think the path from RTR calls the hook function() - findOrCreateObjectiveTemplate. I see it export from objective hooks file but only implemented in the AR path.

Reply:
you are correct, good catch. I have already updated it in the [frontend PR](https://github.com/HHS/Head-Start-TTADP/pull/2716) (still in draft) since I ran into this issue while testing.